contracts/tge/HATAirdrop.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.16;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/MerkleProofUpgradeable.sol";
+import "../tokenlock/ITokenLockFactory.sol";
+
+/*
+An airdrop contract that transfers tokens based on a merkle tree.
+*/
+contract HATAirdrop is OwnableUpgradeable {
+    error CannotRedeemBeforeStartTime();
+    error CannotRedeemAfterDeadline();
+    error LeafAlreadyRedeemed();
+    error InvalidMerkleProof();
+    error CannotRecoverBeforeDeadline();
+
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    bytes32 public root;
+    uint256 public startTime;
+    uint256 public deadline;
+    uint256 public lockEndTime;
```

Comment:
we should consider lock length

Reply:
So here we can either just start from a certain date to a certain date, or we can start on claim and end after X time. I implemented the former but we the latter is also a valid option, we just need to decide what behavior we want here.